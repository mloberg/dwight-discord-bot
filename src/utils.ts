export function between(min: number, max: number): number {
    return Math.floor(Math.random()*(max-min+1)+min);
}

export function env(key: string, _default: string): string {
    return process.env[key] || _default;
}

export function rand<T>(items: T[]): T {
    return items[Math.floor(Math.random() * items.length)];
}

export function ucfirst(input: string): string {
    return input.charAt(0).toUpperCase() + input.slice(1).toLowerCase();
}

export function roll(expr: string): number {
    const input = expr.toLowerCase().match(/^(\d+)?d(\d+)([+-×x])?(\d+)?/);
    if (!input) {
        throw new Error("Invalid roll expression");
    }

    const count = Number(input[1] || 1);
    const dice = Number(input[2]);
    const modifier = input[3];
    const mod = Number(input[4]);

    let total = 0;
    for (let index = 0; index < count; index++) {
        total += Math.floor(Math.random() * (dice - 1 + 1)) + 1;
    }

    if ("+" === modifier) {
        total += mod;
    } else if ("-" === modifier) {
        total -= mod;
    } else if ("x" === modifier || "×" === modifier) {
        total = total * mod;
    }

    return total;
}
