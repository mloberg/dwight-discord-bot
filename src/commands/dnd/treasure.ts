import { Command, CommandMessage } from "discord.js-commando";
import { isString } from "util";

import treasure, { crIndex } from "../../data/treasure";
import { rand, roll as _d } from "../../utils";

export default class Treasure extends Command {
    constructor(client) {
        super(client, {
            name: "treasure",
            aliases: ["loot"],
            group: "dnd",
            memberName: "treasure",
            description: "Generate random treasure.",
            examples: ["treasure ind 2", "treasure hoard 5 72"],
            format: "[ind|individual|group|hoard] CR <d100>",
            args: [
                {
                    key: "type",
                    prompt: "What type of treasure to generate?",
                    type: "string",
                    validate: v => ["ind", "individual", "group", "hoard"].includes(v), // use oneOf in next version
                },
                {
                    key: "cr",
                    label: "challengeRating",
                    prompt: "What was the challenge rating?",
                    type: "integer",
                    min: 0,
                },
                {
                    key: "roll",
                    prompt: "What was the d100 roll?",
                    type: "integer",
                    default: 0,
                    min: 1,
                    max: 100,
                },
            ],
        });
    }

    async run(msg: CommandMessage, args) {
        const roll = 0 === args.roll ? _d("d100") : args.roll;
        const table = (["ind", "individual"].includes(args.type)
            ? treasure.individual
            : treasure.hoard)[crIndex(args.cr)][roll - 1];

        let reply = "You found:";
        for (const [key, type, value] of table) {
            if (!value) {
                reply += `\n* ${_d(key)} ${type}`;
                continue;
            }

            for (let index = 0; index < _d(key); index++) {
                const item = await this.resolveItem(rand(value));
                reply += `\n* ${type}: ${item}`;
            }
        }

        return msg.channel.send(reply);
    }

    private async resolveItem(value: string|Function): Promise<string> {
        return isString(value) ? value : await value();
    }
}
