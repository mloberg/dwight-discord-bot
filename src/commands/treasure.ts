import { isString, sample } from 'lodash';

import { crIndex, hoard, individual } from '../data/treasure';
import { FriendlyError } from '../error';
import { Command } from '../types';
import { roll } from '../utils';

const command: Command = {
    name: 'treasure',
    alias: ['loot'],
    args: /(?<hoard>hoard )?(?<cr>\d+)(?<roll> \d+)?/,
    description: 'Give me the loot!',
    usage: '[hoard] <cr> [d100]',
    examples: ['4', '6 86', 'hoard 10'],
    async run(message, { groups }) {
        const cr = Number(groups.cr);
        const dice = Number(groups.roll || roll('d100'));

        if (isNaN(cr)) {
            throw new FriendlyError('Missing challenge rating');
        }

        const tables = groups.hoard ? hoard : individual;
        const table = tables[crIndex(cr)][dice - 1];

        const reply = ['You found:'];
        for (const record of table) {
            const key = record[0] as string;
            const type = record[1];
            const value = record[2] as (string | (() => Promise<string>))[];

            if (!value) {
                reply.push(`* ${roll(key)} ${type}`);
                continue;
            }

            for (let index = 0; index < roll(key); index++) {
                const item = sample(value) ?? '';
                const resolved = isString(item) ? item : await item();
                reply.push(`* ${type}: ${resolved}`);
            }
        }

        await message.delete();

        return message.channel.send(reply, { split: true });
    },
};

export default command;
