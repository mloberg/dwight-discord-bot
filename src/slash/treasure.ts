import { isString, sample } from 'lodash';

import { crIndex, hoard, individual } from '../data/treasure';
import { SlashCommand } from '../types';
import { roll } from '../utils';

const treasure: SlashCommand = {
    name: 'treasure',
    description: 'Give me the loot!',
    options: [
        {
            name: 'cr',
            description: 'Challenge rating',
            type: 'INTEGER',
            required: true,
        },
        {
            name: 'roll',
            description: 'd100 roll',
            type: 'INTEGER',
        },
        {
            name: 'hoard',
            description: 'Generate a treasure hoard',
            type: 'BOOLEAN',
        },
    ],
    async run(command) {
        const cr = command.options.getInteger('cr', true);
        const dice = command.options.getInteger('roll') || roll('d100');
        const isHoard = command.options.getBoolean('hoard') ?? false;

        const tables = isHoard ? hoard : individual;
        const table = tables[crIndex(cr)][dice - 1];
        const results = [];

        for (const record of table) {
            const [key, type, value] = record;
            const count = roll(key as string);

            if (!value) {
                results.push(`* ${count} ${type}`);
                continue;
            }

            for (let index = 0; index < count; index++) {
                const item = sample(value) ?? '';
                const resolved = isString(item) ? item : await item();
                results.push(`* ${type}: ${resolved}`);
            }
        }

        await command.reply(results.join('\n'));
    },
};

export default treasure;
