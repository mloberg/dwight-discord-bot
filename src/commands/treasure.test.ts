import { CommandInteraction } from 'discord.js';
import { mocked } from 'ts-jest/utils';

import treasure from './treasure';

jest.mock('discord.js', () => ({
    CommandInteraction: jest.fn().mockImplementation(() => ({
        reply: jest.fn(),
        options: {
            getString: jest.fn(),
            getInteger: jest.fn(),
            getBoolean: jest.fn(),
        },
    })),
}));
jest.mock('../data/spells', () => {
    return () => [
        {
            spell: 'Cantrip',
            level: 0,
            school: 'Illusion',
            class: ['Wizard'],
        },
        {
            spell: 'One',
            level: 1,
            school: 'Illusion',
            class: ['Wizard'],
        },
        {
            spell: 'Two',
            level: 2,
            school: 'Illusion',
            class: ['Wizard'],
        },
        {
            spell: 'Three',
            level: 3,
            school: 'Illusion',
            class: ['Wizard'],
        },
        {
            spell: 'Four',
            level: 4,
            school: 'Illusion',
            class: ['Wizard'],
        },
        {
            spell: 'Five',
            level: 5,
            school: 'Illusion',
            class: ['Wizard'],
        },
        {
            spell: 'Six',
            level: 6,
            school: 'Illusion',
            class: ['Wizard'],
        },
        {
            spell: 'Seven',
            level: 7,
            school: 'Illusion',
            class: ['Wizard'],
        },
        {
            spell: 'Eight',
            level: 8,
            school: 'Illusion',
            class: ['Wizard'],
        },
        {
            spell: 'Nine',
            level: 9,
            school: 'Illusion',
            class: ['Wizard'],
        },
    ];
});

describe('/treasure', () => {
    it('is a slash command', () => {
        expect(treasure).toMatchSnapshot();
    });

    it('returns random individual treasure', async () => {
        const command = mocked(new CommandInteraction({} as never, {} as never), true);
        command.options.getInteger.mockReturnValueOnce(1);

        await treasure.handler(command);
        expect(command.reply).toHaveBeenCalledWith(expect.stringMatching(/\* \d+ [csegp]p$/));
    });

    it('returns random individual treasure for a dice roll', async () => {
        const command = mocked(new CommandInteraction({} as never, {} as never), true);
        command.options.getInteger.mockReturnValueOnce(17);
        command.options.getInteger.mockReturnValue(99);

        await treasure.handler(command);
        expect(command.reply).toHaveBeenCalledWith(expect.stringMatching(/\* \d+ [gp]p$/));
    });

    it('returns random treasure hoard', async () => {
        const command = mocked(new CommandInteraction({} as never, {} as never), true);
        command.options.getInteger.mockReturnValueOnce(1);
        command.options.getInteger.mockReturnValueOnce(99);
        command.options.getBoolean.mockReturnValue(true);

        await treasure.handler(command);
        expect(command.reply).toHaveBeenCalledWith(expect.stringMatching(/(Gem|Art|Item): .+$/));
    });
});
