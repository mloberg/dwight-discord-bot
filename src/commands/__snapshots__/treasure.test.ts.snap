// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/treasure is a slash command 1`] = `
Object {
  "config": Object {
    "default_permission": undefined,
    "description": "Give me the loot!",
    "name": "treasure",
    "options": Array [
      Object {
        "choices": undefined,
        "description": "Challenge rating",
        "name": "cr",
        "required": true,
        "type": 4,
      },
      Object {
        "choices": undefined,
        "description": "Value of a d100 roll.",
        "name": "roll",
        "required": false,
        "type": 4,
      },
      Object {
        "description": "Generate a treasure hoard",
        "name": "hoard",
        "required": false,
        "type": 5,
      },
    ],
  },
  "handle": [Function],
}
`;
